File: app/android/app/src/main/java/bosch/smartcampus/thermalcomfortstudy/lib/BandDataAggregate.java
package bosch.smartcampus.thermalcomfortstudy.lib;

import android.util.Log;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by rsukkerd on 6/22/16.
 */
public class BandDataAggregate {
    private static final String LOG_TAG = BandDataAggregate.class.getSimpleName();

    // Raw sensor reading data from band
    private List<Double> rawTemperatureData = new ArrayList<>();
    private List<Long> rawTotalCaloriesData = new ArrayList<>();
    private List<Integer> rawGsrData = new ArrayList<>();
    private List<Float> rawSkinTemperatureData = new ArrayList<>();
    private boolean mHasBandContact = false;

    // Average data
    private List<Double> averageTemperatureData = new ArrayList<>();
    private List<Double> averageCaloriesData = new ArrayList<>();
    private List<Double> averageGsrData = new ArrayList<>();
    private List<Double> averageSkinTemperatureData = new ArrayList<>();

    public BandDataAggregate() {
    }

    /**
     * Put current raw temperature
     * Unit: C
     * Frequency: 1 Hz
     * @pre: BandContactState.WORN
     */
    public void putTemperature(double temperature) {
        if (mHasBandContact) {
            rawTemperatureData.add(temperature);
        }
    }

    /**
     * Put total number of calories since last factory-reset
     * Unit: kCals
     * Frequency: 1 Hz
     * @pre: BandContactState.WORN
     */
    public void putTotalCalories(long totalCalories) {
        if (mHasBandContact) {
            rawTotalCaloriesData.add(totalCalories);
        }
    }

    /**
     * Put current GSR
     * Unit: kOhms
     * Frequency: 0.2 Hz
     * @pre: BandContactState.WORN
     */
    public void putGsr(int resistance) {
        if (mHasBandContact) {
            rawGsrData.add(resistance);
        }
    }

    /**
     * Put current skin temperature (as value changes)
     * Unit: C
     * Frequency: as value changes
     * @pre: BandContactState.WORN
     */
    public void putSkinTemperature(float skinTemperature) {
        if (mHasBandContact) {
            rawSkinTemperatureData.add(skinTemperature);
        }
    }

    public void setBandContactStatus(boolean hasBandContact) {
        mHasBandContact = hasBandContact;
    }

    public boolean hasRawTemperatureData() {
        return !rawTemperatureData.isEmpty();
    }

    public boolean hasRawTotalCaloriesData() {
        return rawTotalCaloriesData.size() >= 2;
    }

    public boolean hasRawGsrData() {
        return !rawGsrData.isEmpty();
    }

    public boolean hasRawSkinTemperatureData() {
        return !rawSkinTemperatureData.isEmpty();
    }

    public void averageTemperature() {
        if (hasRawTemperatureData()) {
            average(rawTemperatureData, averageTemperatureData);
        } else {
            Log.e(LOG_TAG, "No raw temperature data has been received");
        }
    }

    public void averageCalories() {
        if (hasRawTotalCaloriesData()) {
            List<Long> rawCaloriesData = new ArrayList<>();

            for (int i = 0; i < rawTotalCaloriesData.size() - 1; i++) {
                long cur = rawTotalCaloriesData.get(i);
                long next = rawTotalCaloriesData.get(i + 1);
                rawCaloriesData.add(next - cur);
            }

            average(rawCaloriesData, averageCaloriesData);
        } else {
            Log.e(LOG_TAG, "Less than 2 raw total calories data have been received");
        }
    }

    public void averageGsr() {
        if (hasRawGsrData()) {
            average(rawGsrData, averageGsrData);
        } else {
            Log.e(LOG_TAG, "No raw GSR data has been received");
        }
    }

    public void averageSkinTemperature() {
        if (hasRawSkinTemperatureData()) {
            // Skin temperature value is only received when value changes
            float lastRawSkinTemperature = rawSkinTemperatureData.get(rawSkinTemperatureData.size() - 1);
            average(rawSkinTemperatureData, averageSkinTemperatureData);
            rawSkinTemperatureData.add(lastRawSkinTemperature);
        } else {
            Log.e(LOG_TAG, "No raw skin temperature data has been received");
        }
    }

    /**
     * @pre averageTemperatureData.size() > 0
     */
    public double getLastAverageTemperature() {
        return averageTemperatureData.get(averageTemperatureData.size() - 1);
    }

    /**
     * @pre averageCaloriesData.size() > 0
     */
    public double getLastAverageCalories() {
        return averageCaloriesData.get(averageCaloriesData.size() - 1);
    }

    /**
     * @pre averageGsrData.size() > 0
     */
    public double getLastAverageGsr() {
        return averageGsrData.get(averageGsrData.size() - 1);
    }

    /**
     * @pre averageSkinTemperatureData.size() > 0
     */
    public double getLastAverageSkinTemperature() {
        return averageSkinTemperatureData.get(averageSkinTemperatureData.size() - 1);
    }

    private void average(List<? extends Number> rawData, List<Double> averageData) {
        assert  !rawData.isEmpty();
        double average = getAverage(rawData);
        averageData.add(average);
        rawData.clear();
    }

    private double getAverage(List<? extends Number> data) {
        assert !data.isEmpty();
        double sum = 0;
        for (Number temp : data) {
            sum += temp.doubleValue();
        }
        return sum / data.size();
    }
}

File: app/iOS/Thermal Comfort Study/BandDataAggregate.swift
//
//  BandDataAggregate.swift
//  Thermal Comfort Study
//
//  Created by Madhav Achar on 10/14/16.
//  Copyright © 2016 boschsc. All rights reserved.
//

import Foundation
import XCGLogger

class BandDataAggregate: NSObject {
    
    // Raw sensor reading data from band
    private var rawExtTemperatureData = [Double]()
    private var rawTotalCaloriesData = [Double]()
    private var rawGSRData = [Double]()
    private var rawSkinTemperatureData = [Double]()
    private var bandHasSkinContact = false
    
    // Average of sensor reading data
    private var avgExtTemperature: Double?
    private var avgRateOfCaloriesBurned: Double?
    private var avgGSR: Double?
    private var avgSkinTemperature: Double?
    
    //MARK: Save Sensor Data
    /**
     * Put current band connection status (if user is wearing device)
     */
    func setBand(contactStatus: Bool) {
        bandHasSkinContact = contactStatus
    }
    
    /**
     * Put current raw external temperatures
     * Unit: kCals
     * Frequency: 1 Hz
     * @pre: Band has to be worn
     */
    func addExtTemperatureReading(_ temp: Double) {
        if(bandHasSkinContact) {
            rawExtTemperatureData.append(temp)
        }
    }
    
    /**
     * Put total number of calories since last factory-reset
     * Unit: kCals
     * Frequency: 1 Hz
     * @pre: Band has to be worn
     */
    func addTotalCaloriesReading(_ totalCalories: UInt) {
        if(bandHasSkinContact) {
            rawTotalCaloriesData.append(Double(totalCalories))
        }
    }
    
    /**
     * Put current raw gsr
     * Unit: kOhms
     * Frequency: 0.2 Hz
     * @pre: Band has to be worn
     */
    func addGSRReading(_ gsr: UInt) {
        if(bandHasSkinContact) {
            rawGSRData.append(Double(gsr))
        }
    }
    
    /**
     * Put current raw skin temperature
     * Unit: C
     * Frequency: as value changes
     * @pre: Band has to be worn
     */
    func addSkinTemperatureReading(_ temp: Double) {
        if(bandHasSkinContact) {
            rawSkinTemperatureData.append(temp)
        }
    }
    
    // MARK: Sensor Data Checks
    func hasRawExtTemperatureData() -> Bool {
        return !rawExtTemperatureData.isEmpty
    }
    
    func hasRawTotalCaloriesData() -> Bool {
        return rawTotalCaloriesData.count >= 2
    }
    
    func hasRawGSRData() -> Bool {
        return !rawGSRData.isEmpty
    }
    
    func hasRawSkinTemperatureData() -> Bool {
        return !rawSkinTemperatureData.isEmpty
    }
    
    func clearData() {
        if (hasRawGSRData()) {
            rawGSRData.removeAll()
        }
        
        if (hasRawExtTemperatureData()) {
            rawExtTemperatureData.removeAll()
        }
        
        if (hasRawSkinTemperatureData()) {
            rawSkinTemperatureData.removeAll()
        }
        
        if (hasRawTotalCaloriesData()){
            rawTotalCaloriesData.removeAll()
        }
    }
    
    // MARK: Average Sensor Data
    func averageSensorData() {
        averageExtTemperature()
        averageRateOfCaloriesBurned()
        averageGSR()
        averageSkinTemperature()
    }
    
    private func averageExtTemperature() {
        if(self.hasRawExtTemperatureData()) {
            avgExtTemperature = average(rawExtTemperatureData)
            rawExtTemperatureData.removeAll()
        } else {
            avgExtTemperature = nil
            log.warning("No barometer data has been received")
        }
    }
    
    private func averageRateOfCaloriesBurned() {
        if(hasRawTotalCaloriesData()) {
            var rateOfCaloriesData = [Double]()
            
            // Create a rate of calories burned array from diff each reading
            for i in 0..<(rawTotalCaloriesData.count - 1) {
                let curr = rawTotalCaloriesData[i]
                let next = rawTotalCaloriesData[i + 1]
                rateOfCaloriesData.append(next - curr)
            }
            
            avgRateOfCaloriesBurned = average(rateOfCaloriesData)
            rawTotalCaloriesData.removeAll()
        } else {
            avgRateOfCaloriesBurned = nil
            log.warning("No calorie data has been received")
        }
    }
    
    private func averageGSR() {
        if(self.hasRawGSRData()) {
            avgGSR = average(rawGSRData)
            rawGSRData.removeAll()
        } else {
            avgGSR = nil
            log.warning("No GSR data has been received")
        }
    }
    
    private func averageSkinTemperature() {
        if(self.hasRawSkinTemperatureData()) {
            avgSkinTemperature = average(rawSkinTemperatureData)
            rawSkinTemperatureData.removeAll()
        } else {
            avgSkinTemperature = nil
            log.warning("No skin temperature data has been received")
        }
    }
    
    // MARK: Return Average Sensor Data
    func getAverageExtTemperature() -> Double? {
        return avgExtTemperature
    }
    
    func getAverageCalories() -> Double? {
        return avgRateOfCaloriesBurned
    }
    
    func getAverageGSR() -> Double? {
        return avgGSR
    }
    
    func getAverageSkinTemperature() -> Double? {
        return avgSkinTemperature
    }
    
    // MARK: Averaging Helper Functions
    private func average(_ sensorData: [Double]) -> Double {
        var sum = 0.0
        for data in sensorData {
            sum += data
        }
        return sum / Double(sensorData.count)
    }
}
